 # PROYECTO SPRINT 10

### Descripción del proyecto
<p>
Trabajo en la compañía de extracción de petróleo OilyGiant. Mi objetivo es encontrar los mejores lugares donde abrir 200 pozos nuevos de petróleo.

Para completar esta tarea, tendrás que realizar los siguientes pasos:

- Leer los archivos con los parámetros recogidos de pozos petrolíferos en la región seleccionada: calidad de crudo y volumen de reservas.
- Crear un modelo para predecir el volumen de reservas en pozos nuevos.
- Elegir los pozos petrolíferos que tienen los valores estimados más altos.


Elegir la región con el beneficio total más alto para los pozos petrolíferos seleccionados.</P>



## 1. Descarga y prepara los datos. Explica el procedimiento.

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

In [2]:
geo_0 = pd.read_csv('/datasets/geo_data_0.csv')
geo_1 = pd.read_csv('/datasets/geo_data_1.csv')
geo_2 = pd.read_csv('/datasets/geo_data_2.csv')

FileNotFoundError: [Errno 2] No such file or directory: '/datasets/geo_data_0.csv'

In [ ]:
display(geo_0.head())
geo_0.info()

id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [ ]:
display(geo_1.sample(5))
geo_1.info()

id         f0        f1        f2     product
66583  gkskc -16.484378 -1.167003  4.009275  110.992147
92569  B00zA  -3.114295 -4.817075  4.004798  110.992147
50847  hQ1l7  -3.918710 -0.692506  1.002452   30.132364
12856  exD9L   8.031348 -6.629580  2.996916   80.859783
14299  i0uUW  -2.518690 -1.451133  3.999191  110.992147

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [ ]:
display(geo_2.sample(5))
geo_2.info()

id        f0        f1         f2     product
67623  9l7Xe -3.440473  0.161187   2.422281  130.544070
93742  EqNhh  2.353356 -0.222156   0.056869   69.480931
51033  U8Gf6  2.188667 -2.311510   2.698711  152.963081
95422  9AfGH -1.034705  3.126783  10.443381   63.039245
35228  qFxwV  0.915502 -1.122161   5.194884   81.601721

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [ ]:
# Verificar valores nulos
print("Valores nulos en geo_0:\n", geo_0.isnull().sum())
print("\nValores nulos en geo_1:\n", geo_1.isnull().sum())
print("\nValores nulos en geo_2:\n", geo_2.isnull().sum())

Valores nulos en geo_0:
 id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Valores nulos en geo_1:
 id         0
f0         0
f1         0
f2         0
product    0
dtype: int64

Valores nulos en geo_2:
 id         0
f0         0
f1         0
f2         0
product    0
dtype: int64


#### No hay valores nulos en ninguno de los conjuntos de datos, así que podemos proceder con la división del conjunto de datos.

## 2. Entrena y prueba el modelo para cada región en geo_data_0.csv:

In [ ]:
#Entrenar y probar el modelo para cada región en geo_data_0
def train_test_model(region_data):
    #Dividir los datos en entrenamiento y validación
    features = region_data.drop(columns=['id','product'])
    target = region_data['product']
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size = 0.25, random_state = 12345)

    #Entrenar el Modelo
    model = LinearRegression()
    model.fit(features_train, target_train)

    #Realizar predicciones para el conjunto de validación
    predicted_valid = model.predict(features_valid)

    #Guardar las predicciones y las respuestas correctas 
    predictions_df = pd.DataFrame({'Predicted': predicted_valid, 'Actual': target_valid})

    #Mostrar el volumen medio de reservas predicho y RMSE del modelo
    avg_predicted_volume = predicted_valid.mean()
    rmse = (mean_squared_error(target_valid, predicted_valid))**0.5
 
    return predictions_df, features_valid, avg_predicted_volume, rmse

# Entrenar y probar el modelo para cada región
predictions_0, targets_0, mean_predicted_volume_0, rmse_0 = train_test_model(geo_0)
predictions_1, targets_1, mean_predicted_volume_1, rmse_1 = train_test_model(geo_1)
predictions_2, targets_2, mean_predicted_volume_2, rmse_2 = train_test_model(geo_2)

print("Region 0 - Mean Predicted Volume:", mean_predicted_volume_0, "RMSE:", rmse_0)
print("Region 1 - Mean Predicted Volume:", mean_predicted_volume_1, "RMSE:", rmse_1)
print("Region 2 - Mean Predicted Volume:", mean_predicted_volume_2, "RMSE:", rmse_2)


Region 0 - Mean Predicted Volume: 92.59256778438035 RMSE: 37.5794217150813
Region 1 - Mean Predicted Volume: 68.72854689544602 RMSE: 0.8930992867756165
Region 2 - Mean Predicted Volume: 94.96504596800489 RMSE: 40.02970873393434


 #### La región 0 muestra una alta variabilidad en las predicciones en comparación con las otras dos regiones, como se evidencia por su alto RMSE. Esto podría indicar una mayor complejidad o variabilidad en los datos de esa región.

 #### La región 1 tiene el RMSE más bajo (0.89), lo que indica que las predicciones están muy cerca de los valores reales. Por otro lado, la región 0 tiene el RMSE más alto (37.58), lo que sugiere que las predicciones son menos precisas en esa región. La región 2 está en algún punto intermedio en términos de precisión de predicción (RMSE: 40.03).

## 3. Prepárate para el cálculo de ganancias:

In [ ]:
# Almacenar valores necesarios
investment_per_well = 100000000 / 200  # Inversión por pozo
threshold_value = 500000 / 4.5  # Valor mínimo de reservas en miles de barriles

# Comparar con el umbral de pérdidas
print("Umbral de pérdidas (en miles de barriles):", threshold_value)

# Comparar con la cantidad media de reservas en cada región
print("\nComparación con la cantidad media de reservas en cada región:")
print("Región 0 - Media de reservas:", geo_0['product'].mean())
print("Región 1 - Media de reservas:", geo_1['product'].mean())
print("Región 2 - Media de reservas:", geo_2['product'].mean())

Umbral de pérdidas (en miles de barriles): 111111.11111111111

Comparación con la cantidad media de reservas en cada región:
Región 0 - Media de reservas: 92.50000000000001
Región 1 - Media de reservas: 68.82500000000002
Región 2 - Media de reservas: 95.00000000000004


## 4. Función para calcular la ganancia de un conjunto de pozos seleccionados

In [ ]:
def calculate_profit(predictions_df, investment_per_well):
    # Ordenar las predicciones por valor descendente y seleccionar las primeras 200
    top_200_wells = predictions_df.sort_values(by='Predicted', ascending=False).head(200)

    # Calcular el volumen objetivo de reservas
    target_volume = top_200_wells['Actual'].sum()

    # Calcular la ganancia potencial
    profit = target_volume * 4.5 * 1000 - investment_per_well * 200

    return profit

# Calcular ganancia potencial para cada región
profit_0 = calculate_profit(predictions_0, investment_per_well)
profit_1 = calculate_profit(predictions_1, investment_per_well)
profit_2 = calculate_profit(predictions_2, investment_per_well)

print("Ganancia potencial para la Región 0:", profit_0)
print("Ganancia potencial para la Región 1:", profit_1)
print("Ganancia potencial para la Región 2:", profit_2)

Ganancia potencial para la Región 0: 33208260.431398526
Ganancia potencial para la Región 1: 24150866.966815114
Ganancia potencial para la Región 2: 27103499.635998324


#### La región 0 tiene el mayor potencial de ganancia, pero también puede tener un mayor riesgo debido a la mayor variabilidad en las predicciones. Las regiones 1 y 2 pueden ser consideradas como alternativas con menores ganancias potenciales pero posiblemente con menor riesgo. La elección entre las regiones dependerá de la tolerancia al riesgo y las preferencias específicas del negocio.

 ## 5. Calcular riesgos y ganancias para cada región

In [ ]:
def calculate_bootstrap_profit(predictions_df, investment_per_well, num_samples=1000):
    # Inicializar lista para almacenar los beneficios de cada muestra
    bootstrap_profits = []

    # Realizar bootstrapping
    for _ in range(num_samples):
        # Seleccionar una muestra aleatoria con reemplazo del mismo tamaño que el conjunto de predicciones
        sample = predictions_df.sample(n=500, replace=True)

        # Calcular el beneficio para la muestra seleccionada
        profit = calculate_profit(sample, investment_per_well)

        # Agregar el beneficio a la lista
        bootstrap_profits.append(profit)

    return bootstrap_profits

# Calcular beneficios mediante bootstrapping para cada región
bootstrap_profits_0 = calculate_bootstrap_profit(predictions_0, investment_per_well)
bootstrap_profits_1 = calculate_bootstrap_profit(predictions_1, investment_per_well)
bootstrap_profits_2 = calculate_bootstrap_profit(predictions_2, investment_per_well)

# Calcular estadísticas de los beneficios para cada región
mean_profit_0 = np.mean(bootstrap_profits_0)
mean_profit_1 = np.mean(bootstrap_profits_1)
mean_profit_2 = np.mean(bootstrap_profits_2)

ci_95_0 = np.percentile(bootstrap_profits_0, [2.5, 97.5])
ci_95_1 = np.percentile(bootstrap_profits_1, [2.5, 97.5])
ci_95_2 = np.percentile(bootstrap_profits_2, [2.5, 97.5])

# Calcular riesgo de pérdidas
risk_loss_0 = np.mean([1 for profit in bootstrap_profits_0 if profit < 0])
risk_loss_1 = np.mean([1 for profit in bootstrap_profits_1 if profit < 0])
risk_loss_2 = np.mean([1 for profit in bootstrap_profits_2 if profit < 0])

# Expresar riesgo de pérdidas como un porcentaje
risk_loss_0_percentage = (risk_loss_0 / len(bootstrap_profits_0)) * 100
risk_loss_1_percentage = (risk_loss_1 / len(bootstrap_profits_1)) * 100
risk_loss_2_percentage = (risk_loss_2 / len(bootstrap_profits_2)) * 100

# Tratar los NaN como riesgo de pérdidas del 100%
risk_loss_0_percentage = 100 if np.isnan(risk_loss_0_percentage) else risk_loss_0_percentage
risk_loss_1_percentage = 100 if np.isnan(risk_loss_1_percentage) else risk_loss_1_percentage
risk_loss_2_percentage = 100 if np.isnan(risk_loss_2_percentage) else risk_loss_2_percentage

# Mantener solo las regiones con un riesgo de pérdidas inferior al 2.5%
eligible_regions = []
if risk_loss_0_percentage < 2.5:
    eligible_regions.append((mean_profit_0, "Región 0"))
if risk_loss_1_percentage < 2.5:
    eligible_regions.append((mean_profit_1, "Región 1"))
if risk_loss_2_percentage < 2.5:
    eligible_regions.append((mean_profit_2, "Región 2"))

# Seleccionar la región con el beneficio promedio más alto
if eligible_regions:
    max_avg_profit_region = max(eligible_regions)
    print("Se selecciona la", max_avg_profit_region[1], "con un beneficio promedio de", max_avg_profit_region[0])
else:
    print("Ninguna región cumple con el criterio de riesgo de pérdidas inferior al 2.5%")

Se selecciona la Región 1 con un beneficio promedio de 4608094.88606262


#### La Región 1 es la más apropiada para el desarrollo de pozos petrolíferos con un beneficio promedio de : 4608094.8. La elección del pozo no coincide con el punto 4.3 porque en ese punto la elecciónfue la región 0.